From 0a266e11013245868e2a73f24692afaa47b22a16 Mon Sep 17 00:00:00 2001
From: Simon Pichugin <spichugi@redhat.com>
Date: Thu, 16 Oct 2025 22:00:13 -0700
Subject: [PATCH] Issue 7047 - MemberOf plugin logs null attribute name on
 fixup task completion (#7048)

Description: The MemberOf plugin logged "(null)" instead of the attribute
name when the global fixup task completed. This occurred because the config
structure containing the attribute name was freed before the completion log
message was written.

This fix moves the memberof_free_config() call to after the log statement,
ensuring the attribute name is available for logging.

Additionally, the test_shutdown_on_deferred_memberof test has been improved
to properly verify the fixup task behavior by checking that both the "started"
and "finished" log messages contain the correct attribute name.

Fixes: https://github.com/389ds/389-ds-base/issues/7047

Reviewed by: @tbordaz (Thanks!)
---
 .../suites/memberof_plugin/regression_test.py | 21 +++++++++++++++++--
 ldap/servers/plugins/memberof/memberof.c      |  2 +-
 2 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/dirsrvtests/tests/suites/memberof_plugin/regression_test.py b/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
index c6c03f683..ce098615b 100644
--- a/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
+++ b/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
@@ -1425,8 +1425,25 @@ def test_shutdown_on_deferred_memberof(topology_st, request):
     value = memberof.get_memberofneedfixup()
     assert ((str(value).lower() == "yes") or (str(value).lower() == "true"))
 
-    # step 14. fixup task was not launched because by default launch_fixup is no
-    assert len(errlog.match('.*It is recommended to launch memberof fixup task.*')) == 0
+    # step 14. Verify the global fixup started/finished messages
+    attribute_name = 'memberOf'
+    started_lines = errlog.match('.*Memberof plugin started the global fixup task for attribute .*')
+    assert len(started_lines) >= 1
+    for line in started_lines:
+        log.info(f'Started line: {line}')
+        assert f'attribute {attribute_name}' in line
+
+    # Wait for finished messages to appear, then verify no nulls are present
+    finished_lines = []
+    for _ in range(60):
+        finished_lines = errlog.match('.*Memberof plugin finished the global fixup task.*')
+        if finished_lines:
+            break
+        time.sleep(1)
+    assert len(finished_lines) >= 1
+    for line in finished_lines:
+        log.info(f'Finished line: {line}')
+        assert '(null)' not in line
 
     # Check that users memberof and group members are in sync.
     time.sleep(delay)
diff --git a/ldap/servers/plugins/memberof/memberof.c b/ldap/servers/plugins/memberof/memberof.c
index 7b3f55342..4ebd5e838 100644
--- a/ldap/servers/plugins/memberof/memberof.c
+++ b/ldap/servers/plugins/memberof/memberof.c
@@ -934,9 +934,9 @@ perform_needed_fixup(void)
     slapi_ch_free_string(&cookie);
     slapi_ch_free_string(&td.bind_dn);
     slapi_ch_free_string(&td.filter_str);
-    memberof_free_config(&config);
     slapi_log_err(SLAPI_LOG_INFO, MEMBEROF_PLUGIN_SUBSYSTEM,
                   "Memberof plugin finished the global fixup task for attribute %s\n", config.memberof_attr);
+    memberof_free_config(&config);
     return rc;
 }
 
-- 
2.52.0

